README.md
# School-Code-Compare

Compare files containing source code (or any plain text) on similarity to each other.
This may help you to find files which contain similar code.

The approach is simplistic:
Whitespace and comments are trimmed, then all files are compared using the Levenshtein algorithm.

Future releases may bring more sophisticated techniques.


Changes
Revision history for School-Code-Compare
 
{{$NEXT}}
 - switch repo from github to gitlab in dist.ini meta information
 - update comment about © date in script
 - update documentation / fix typos and spelling

0.007     2019-01-17 13:54:55+01:00 Europe/Zurich
 - throw "skipped" comparisons away
 - mention html in help as option for output format

0.006     2017-12-11 16:35:04+01:00 Europe/Zurich
 - support multiline comments, probably buggy ( /* )
 - add 'java' as option, added some code_examples
 - deal with \r (carriage return) at end of file paths (windows)
 - remove all current files from folder doc (thus whole folder)
 - add some documentation for windows

0.005     2017-12-04 16:02:32+01:00 Europe/Zurich
 - auto generated META.json using dzil plugin [MetaJSON] (MANWAR)
 - explode dzil [@Basic] so that [GatherDir] excludes repo license
 - fix minor Kwalitee stuff (add use strict and warnings)
 - fix some minor typos in documentation

0.004     2017-11-30 09:55:14+01:00 Europe/Zurich
 - replace IO::Prompter with IO::Prompt::Simple (less issues on windows)
 - fix issue with user-interaction while reading from STDIN

0.003     2017-11-30 09:01:25+01:00 Europe/Zurich
 - documentation update: function docs for main lib
 - allow absolute paths for Template. System installation seems requiring this

0.002     2017-11-29 10:24:00+01:00 Europe/Zurich
 - documentation update, mainly for the script in bin directory

0.001     2017-11-27 22:25:07+01:00 Europe/Zurich (TRIAL RELEASE)
 - initial release


find_similar_code.pl
# Autor: Boris Däppen, 2015-2016
# No guarantee given, use at own risk and will

use strict;
use warnings;
use utf8;
use feature 'say';

use Text::Levenshtein qw(distance);

# Kombinatorisches Verhalten
# -----------------------------------------------------------------------------
#
# Anzahl Vergleiche:
# n! / ((n - m)! * m!)
# wenn
# n = Anzahl Elemente   (Anzahl Code-Dateien die zu Vergleichen sind)
# m = gezogene Elemente (immer 2, da zwei Dateien miteinander verglichen werden)
#
# Bei 100 Skripte gibt das
# 100! / (98! * 2!) = 4950
#
# Rechner: http://de.numberempire.com/combinatorialcalculator.php

unless ( defined $ARGV[0] ) {
    say "Please define Programming Language";
    exit 1;
}
my $lang = $ARGV[0];

$| = 1;

my $CHARDIFF = 70;

my @files = ();

foreach my $filepath ( <STDIN> ) {
    chomp( $filepath );
    #say "adding '$filepath' ...";

    push @files, $filepath;
}

say '# comparing ' . @files . ' files';
say "#edits\tratio\tlength\tfile1\tfile2";

for (my $i=0; $i < @files - 1; $i++) {
    for (my $j=$i+1; $j < @files; $j++) {

        my ($cleaned_code1, $cleaned_code2);

        if ($lang eq 'python') {
            ($cleaned_code1,
             $cleaned_code2) = prepare_python( $files[$i],  $files[$j] );
        }
        if ($lang eq 'php') {
            ($cleaned_code1,
             $cleaned_code2) = prepare_php   ( $files[$i],  $files[$j] );
        }
        if ($lang eq 'html') {
            ($cleaned_code1,
             $cleaned_code2) = prepare_html  ( $files[$i],  $files[$j] );
        }

        my ($res, $prop, $diff) = measure( $cleaned_code1,  $cleaned_code2);
        say "$res\t$prop\t$diff\t$files[$i]\t$files[$j]";
    }
}

sub prepare_python {
    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";
     
    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";
    
    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ /^#/);
      $row = $1 if ($row =~ /(.*)#.*/);
      $str1 .= $row
    }
    close $fh1;
    
    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ /^#/);
      $row = $1 if ($row =~ /(.*)#.*/);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    #say $str1;
    #say $str2;

    return ($str1, $str2);
}

sub prepare_php {
    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";
     
    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";
    
    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ m!^/!);
      $row = $1 if ($row =~ m!(.*)//.*!);
      $row = $1 if ($row =~ m!(.*)/\*.*!);
      $str1 .= $row
    }
    close $fh1;
    
    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ m!^/!);
      $row = $1 if ($row =~ m!(.*)//.*!);
      $row = $1 if ($row =~ m!(.*)/\*.*!);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    return ($str1, $str2);
}

sub prepare_html {
    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";
     
    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";
    
    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ m/^<!--/);
      $row = $1 if ($row =~ m/(.*)<!--.*/);
      $str1 .= $row
    }
    close $fh1;
    
    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ m/^<!--/);
      $row = $1 if ($row =~ m/(.*)<!--.*/);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    return ($str1, $str2);
}

sub measure {
    my $str1 = shift;
    my $str2 = shift;

    my $diff = length($str1) - length($str2);
    
    $diff = $diff * -1 if ($diff < 0);

    if ($diff > $CHARDIFF) {
        return (-1, -1, $diff);
    }
    else {
        my $distance = distance($str1, $str2);

        my $total_chars = length($str1) + length($str2);
        my $proportion_chars_changes = int(($distance / ($total_chars / 2))*100);

        return ($distance, $proportion_chars_changes, $diff);
    }
}


find_similar_python.pl
use strict;
use warnings;
use utf8;
use feature 'say';

use Text::Levenshtein qw(distance);

unless ( defined $ARGV[0] ) {
    say "Please define Programming Language";
    exit 1;
}
my $lang = $ARGV[0];

$| = 1;

my $CHARDIFF = 70;

my @files = ();

foreach my $filepath ( <STDIN> ) {
    chomp( $filepath );
    #say "adding '$filepath' ...";

    push @files, $filepath;
}

say '# comparing ' . @files . ' files';
say "#edits\tlength\tfile1\tfile2";

for (my $i=0; $i < @files - 1; $i++) {
    for (my $j=$i+1; $j < @files; $j++) {

        my ($cleaned_code1, $cleaned_code2);

        if ($lang eq 'python') {
            ($cleaned_code1,
             $cleaned_code2) = prepare_python( $files[$i],  $files[$j] );
        }
        if ($lang eq 'php') {
            ($cleaned_code1,
             $cleaned_code2) = prepare_php   ( $files[$i],  $files[$j] );
        }
        if ($lang eq 'html') {
            ($cleaned_code1,
             $cleaned_code2) = prepare_html  ( $files[$i],  $files[$j] );
        }

        my ($res, $diff) = measure( $cleaned_code1,  $cleaned_code2);
        say "$res\t$diff\t$files[$i]\t$files[$j]";
    }
}

sub prepare_python {
    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";
     
    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";
    
    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ /^#/);
      $row = $1 if ($row =~ /(.*)#.*/);
      $str1 .= $row
    }
    close $fh1;
    
    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ /^#/);
      $row = $1 if ($row =~ /(.*)#.*/);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    #say $str1;
    #say $str2;

    return ($str1, $str2);
}

sub prepare_php {
    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";
     
    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";
    
    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ m!^/!);
      $row = $1 if ($row =~ m!(.*)//.*!);
      $row = $1 if ($row =~ m!(.*)/\*.*!);
      $str1 .= $row
    }
    close $fh1;
    
    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ m!^/!);
      $row = $1 if ($row =~ m!(.*)//.*!);
      $row = $1 if ($row =~ m!(.*)/\*.*!);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    return ($str1, $str2);
}

sub prepare_html {
    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";
     
    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";
    
    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ m/^<!--/);
      $row = $1 if ($row =~ m/(.*)<!--.*/);
      $str1 .= $row
    }
    close $fh1;
    
    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ m/^<!--/);
      $row = $1 if ($row =~ m/(.*)<!--.*/);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    return ($str1, $str2);
}

sub measure {
    my $str1 = shift;
    my $str2 = shift;

    my $diff = length($str1) - length($str2);
    
    $diff = $diff * -1 if ($diff < 0);

    if ($diff > $CHARDIFF) {
        return (-1, $diff);
    }
    else {
        return (distance($str1, $str2), $diff);
    }
}


lib/School/Code/Simplify.pm
package School::Code::Simplify;

sub new {
    my $class = shift;

    my $self = {
               };
    bless $self, $class;

    return $self;
}

sub hashy {
    my $self = shift;

    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";

    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";

    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ /^#/);
      $row = $1 if ($row =~ /(.*)#.*/);
      $str1 .= $row
    }
    close $fh1;

    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ /^#/);
      $row = $1 if ($row =~ /(.*)#.*/);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    return ($str1, $str2);
}

sub slashy {
    my $self = shift;

    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";

    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";

    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ m!^/!);
      $row = $1 if ($row =~ m!(.*)//.*!);
      $row = $1 if ($row =~ m!(.*)/\*.*!);
      $str1 .= $row
    }
    close $fh1;

    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ m!^/!);
      $row = $1 if ($row =~ m!(.*)//.*!);
      $row = $1 if ($row =~ m!(.*)/\*.*!);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    return ($str1, $str2);
}

sub html {
    my $self = shift;

    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";

    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";

    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      next if ($row =~ m/^<!--/);
      $row = $1 if ($row =~ m/(.*)<!--.*/);
      $str1 .= $row
    }
    close $fh1;

    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      next if ($row =~ m/^<!--/);
      $row = $1 if ($row =~ m/(.*)<!--.*/);
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    return ($str1, $str2);
}

sub txt {
    my $self = shift;

    my $f1 = shift;
    my $f2 = shift;

    open(my $fh1, '<:encoding(UTF-8)', $f1)
      or die "Could not open file '$f1' $!";

    open(my $fh2, '<:encoding(UTF-8)', $f2)
      or die "Could not open file '$f2' $!";

    my $str1 = '';
    while (my $row = <$fh1>) {
      chomp $row;
      $str1 .= $row
    }
    close $fh1;

    my $str2 = '';
    while (my $row = <$fh2>) {
      chomp $row;
      $str2 .= $row
    }
    close $fh2;

    # Whitespace raus
    $str1 =~ s/\s*//g;
    $str2 =~ s/\s*//g;

    return ($str1, $str2);
}

1;


lib/School/Code/Compare/Out.pm
package School::Code::Compare::Out;
# ABSTRACT: organize the output to different formats

use strict;
use warnings;

use Template;
use School::Code::Compare::Out::Template::Path;

sub new {
    my $class = shift;

    my $self = {
                    name   => time(),
                    format => 'tab',
                    lines  => [],
                    title  => 'Comparion',
                    description => 'a list of files',
               };
    bless $self, $class;

    return $self;
}

sub set_name {
    my $self = shift;

    $self->{name} = shift;

    return $self;
}

sub set_format {
    my $self = shift;

    $self->{format} = shift;

    return $self;
}

sub set_lines {
    my $self = shift;

    $self->{lines} = shift;

    return $self;
}

sub set_title {
    my $self = shift;

    $self->{title} = shift;

    return $self;
}

sub set_description {
    my $self = shift;

    $self->{description} = shift;

    return $self;
}

sub write {
    my $self        = shift;

    my @result      = @{$self->{lines}};
    my $format      =   $self->{format};
    my $filename    =   $self->{name};
    my $title       =   $self->{title};
    my $description =   $self->{description};

    my $tt     = Template->new( ABSOLUTE => 1 );
    my $tt_dir = School::Code::Compare::Out::Template::Path->get();
    
    # sort by ratio, but make sure undef values are "big" (meaning, bottom/last)
    my @result_sorted = sort { return  1 if (not defined $a->{ratio});
                               return -1 if (not defined $b->{ratio});
                               return $a->{ratio} <=> $b->{ratio};
                             } @result;
    
    # we render all rows, appending it to one string
    my $rendered_data_rows = '';
    
    foreach my $comparison (@result_sorted) {
        my $vars = {
            ratio        => $comparison->{ratio},
            distance     => $comparison->{distance},
            delta_length => $comparison->{delta_length},
            suspicious   => $comparison->{suspicious},
            file1        => $comparison->{file1},
            file2        => $comparison->{file2},
            comment      => $comparison->{comment},
        };
    
        $tt->process("$tt_dir/$format.tt", $vars, \$rendered_data_rows)
            || die $tt->error(), "\n";
    }
    
    # render again, this time merging the rendered rows into the wrapping body
    $tt->process(   "$tt_dir/Body$format.tt",
                    {
                      data        => $rendered_data_rows,
                      title       => $title,
                      description => $description
                    },
                    $filename
                )   || die $tt->error(), "\n";

}

1;


lib/School/Code/Compare/Judge.pm
package School::Code::Compare::Judge;
# ABSTRACT: guess if two strings are so similar, that it's maybe cheating

use strict;
use warnings;


sub new {
    my $class = shift;

    my $self = {
                    suspicious_ratio        => 40,
                    highly_suspicious_ratio => 20,
               };
    bless $self, $class;

    return $self;
}

sub set_suspicious_ratio {
    my $self = shift;

    $self->{suspicious_ratio} = shift;

    return $self;
}

sub set_highly_suspicious_ratio {
    my $self = shift;

    $self->{highly_suspicious_ratio} = shift;

    return $self;
}

sub look {
    my $self       = shift;
    my $comparison = shift;

    $comparison->{suspicious}        = 0;
    $comparison->{highly_suspicious} = 0;

    return () unless (defined $comparison->{ratio});

    if ($comparison->{ratio} <= $self->{suspicious_ratio}) {
        $comparison->{suspicious} = 1;
        if ($comparison->{ratio} <= $self->{highly_suspicious_ratio}) {
            $comparison->{suspicious} = 2;
        }
    }
}

1;

__END__

=encoding utf8

=head1 SYNOPSIS

 use School::Code::Compare::Judge;

 my $judge  = School::Code::Compare::Judge->new();

 # this will alter the content of the hash argument provided
 $judge->look($comparison);
 
=head1 FUNCTIONS

=head2 set_suspicious_ratio

=head2 set_highly_suspicious_ratio

=head2 look


lib/School/Code/Simplify/Comments.pm
package School::Code::Simplify::Comments;
# ABSTRACT: roughly trim whitespace, depending on commenting style

use strict;
use warnings;

sub new {
    my $class = shift;

    my $self = {
               };
    bless $self, $class;

    return $self;
}

sub hashy {
    my $self      = shift;
    my $lines_ref = shift;

    my $visibles      = '';
    my @signes_lines  = ();
    my $signes        = '';

    foreach my $row (@{$lines_ref}) {
      next if ($row =~ /^#/);
      $row = $1 if ($row =~ /(.*)#.*/);
      $row =~ s/\s*//g;
      next if ($row eq '');
      $visibles .= $row;
      $row =~ s/[a-zA-Z0-9]//g;
      $signes .= $row;
      push @signes_lines, $row;
    }

    my $sorted_sortedlines = join '', sort { $a cmp $b } @signes_lines;

    return  {
                visibles => $visibles,
                signes   => $signes,
                signes_ordered => $sorted_sortedlines,
            };
}

sub slashy {
    my $self      = shift;
    my $lines_ref = shift;

    my $visibles      = '';
    my @signes_lines  = ();
    my $signes        = '';

    foreach my $row (@{$lines_ref}) {
      next if ($row =~ m!^/!);
      next if ($row =~ m!^\s*\*!);
      $row = $1 if ($row =~ m!(.*)//.*!);
      $row = $1 if ($row =~ m!(.*)/\*.*!);
      $row =~ s/\s*//g;
      next if ($row eq '');
      $visibles .= $row;
      $row =~ s/[a-zA-Z0-9]//g;
      $signes .= $row;
      push @signes_lines, $row;
    }

    my $sorted_sortedlines = join '', sort { $a cmp $b } @signes_lines;

    return  {
                visibles => $visibles,
                signes   => $signes,
                signes_ordered => $sorted_sortedlines,
            };
}

sub html {
    my $self      = shift;
    my $lines_ref = shift;

    my $visibles      = '';
    my @signes_lines  = ();
    my $signes        = '';

    foreach my $row (@{$lines_ref}) {
      next if ($row =~ m/^<!--/);
      $row = $1 if ($row =~ m/(.*)<!--.*/);
      $row =~ s/\s*//g;
      next if ($row eq '');
      $visibles .= $row;
      $row =~ s/[a-zA-Z0-9]//g;
      $signes .= $row;
      push @signes_lines, $row;
    }

    my $sorted_sortedlines = join '', sort { $a cmp $b } @signes_lines;

    return  {
                visibles => $visibles,
                signes   => $signes,
                signes_ordered => $sorted_sortedlines,
            };
}

sub txt {
    my $self      = shift;
    my $lines_ref = shift;

    my $visibles      = '';
    my @signes_lines  = ();
    my $signes        = '';

    foreach my $row (@{$lines_ref}) {
      $row =~ s/\s*//g;
      next if ($row eq '');
      $visibles .= $row;
      $row =~ s/[a-zA-Z0-9]//g;
      $signes .= $row;
      push @signes_lines, $row;
    }

    my $sorted_sortedlines = join '', sort { $a cmp $b } @signes_lines;

    return  {
                visibles => $visibles,
                signes   => $signes,
                signes_ordered => $sorted_sortedlines,
            };
}

1;


lib/School/Code/Compare.pm
package School::Code::Compare;
# ABSTRACT: 'naive' metrics for code similarity

use strict;
use warnings;

use Text::Levenshtein::XS qw(distance);

sub new {
    my $class = shift;

    my $self = {
                    max_char_diff  => 70,
                    min_char_total => 20,
                    max_distance   => 300,
               };
    bless $self, $class;

    return $self;
}

sub set_max_char_difference {
    my $self = shift;

    $self->{max_char_diff} = shift;

    # make this chainable in OO-interface
    return $self;
}

sub set_min_char_total {
    my $self = shift;

    $self->{min_char_total} = shift;

    # make this chainable in OO-interface
    return $self;
}

sub set_max_distance {
    my $self = shift;

    $self->{max_distance} = shift;

    # make this chainable in OO-interface
    return $self;
}

sub measure {
    my $self = shift;

    my $str1 = shift;
    my $str2 = shift;

    my $length_str1 = length($str1);
    my $length_str2 = length($str2);


    if ($self->{min_char_total} > $length_str1
     or $self->{min_char_total} > $length_str2) {
        return {
            distance     => undef,
            ratio        => undef,
            delta_length => undef,
            comment      => 'skipped: smaller than '
                            . $self->{min_char_total},
        };
    }

    my $diff = $length_str1 - $length_str2;

    $diff = $diff * -1 if ($diff < 0);

    if ($diff > $self->{max_char_diff}) {
        return {
            distance     => undef,
            ratio        => undef,
            delta_length => $diff,
            comment      => 'skipped: delta in length bigger than '
                            . $self->{max_char_diff},
        };
    }
    else {
        my $distance = distance($str1, $str2, $self->{max_distance});

        if (defined $distance) {

            my $total_chars = $length_str1 + $length_str2;
            my $proportion_chars_changes =
                                    int(($distance / ($total_chars / 2))*100);

            return {
                distance     => $distance,
                ratio        => $proportion_chars_changes,
                delta_length => $diff,
                comment      => 'comparison done',
            };
        }
        else {
            return {
                distance     => undef,
                ratio        => undef,
                delta_length => $diff,
                comment      => 'skipped: distance higher than '
                                . $self->{max_distance},
            };
        }
    }
}

1;

__END__

=encoding utf8

=head1 SYNOPSIS

This distribution ships a script.
You might want to look at the script L<compare-code> in the C<bin> directory.
For documentation of the used libraries, keep on reading.

This calculates the Levenshtein Difference for two files, if they meet certain criterias:

 use School::Code::Compare;

 my $comparer   = School::Code::Compare->new()                                    
                                       ->set_max_char_difference(400)             
                                       ->set_min_char_total     ( 20)             
                                       ->set_max_distance       (400);
 
 my $comparison = $comparer->measure( 'use strict; print "Hello\n";',
                                      'use v5.22; say "Hello";'
                                    ); 

 print $comparison->{distance} if $comparison   # 13

=head1 FUNCTIONS

=head2 set_max_char_difference

Don't even start comparison, if the difference in char count is higher than set.

=head2 set_min_char_total

Don't even start comparison if a file is below this char count.

=head2 set_max_distance

Abort comparison (in the midst of comparison), if distance is becoming higher then set value.

=head2 measure

Do a comparison for two strings.
Gives back a hash reference with different information:

 # (example output from synopsis)
 {
   'distance'     => 13,
   'ratio'        => 50,
   'comment'      => 'comparison done',
   'delta_length' => 5
 };

=over 4

=item distance

The Levenshtein Distance.
See L<Text::Levenshtein::XS> for more information.

=item ratio

The ratio of the distance in chars to the average length of the compared strings.
A ratio of zero means, the strings are similar.
A ratio of 50 means, that 50% of a string is different.

My experience is, that if you get a ratio below 30% you have to start looking if the code was copied and altered (if your concern is to find 'cheaters' in educational/school environments).
This method of measurement is by no means well established.
It may be even 'naive', but it just seems to work out quite well.
See L<School::Code::Compare::Judge> to see, how the results are currently interpreted.

=item comment

A comment on how the comparison went.

=item delta_length

Difference in length (chars) of the two compared strings.

=back
